From 225fc4d7be5f76216203fb4b8e5cbf3b72df1037 Mon Sep 17 00:00:00 2001
From: saranya <saranya.gopal@intel.com>
Date: Thu, 30 May 2019 22:06:00 +0530
Subject: [PATCH] PowerManagerService: Switch to USB host mode before suspend

This patch writes to usb role sysfs interface to
switch the dual role port to host mode before suspend.
It will be out back to device mode on resume.

Change-Id: I3c38d856e2e2ca00ba1575fcfe0d38bbf9fe6683
Signed-off-by: saranya <saranya.gopal@intel.com>
---
 .../server/power/PowerManagerService.java     | 25 ++++++++++++++-----
 1 file changed, 19 insertions(+), 6 deletions(-)

diff --git a/services/core/java/com/android/server/power/PowerManagerService.java b/services/core/java/com/android/server/power/PowerManagerService.java
index ab138e3e468..f434bf012ee 100644
--- a/services/core/java/com/android/server/power/PowerManagerService.java
+++ b/services/core/java/com/android/server/power/PowerManagerService.java
@@ -119,6 +119,7 @@ public final class PowerManagerService extends SystemService
         implements Watchdog.Monitor {
     private static final String TAG = "PowerManagerService";
     private static final String USB_CONFIG_PROPERTY = "sys.usb.config";
+    private static final String ROLE_SYSFS = "/sys/class/usb_role/intel_xhci_usb_sw-role-switch/role";
 
     private static final boolean DEBUG = false;
     private static final boolean DEBUG_SPEW = DEBUG && true;
@@ -1433,13 +1434,11 @@ public final class PowerManagerService extends SystemService
                     Slog.i(TAG, "Going to sleep by application request (uid " + uid +")...");
                     reason = PowerManager.GO_TO_SLEEP_REASON_APPLICATION;
 		    SystemProperties.set(USB_CONFIG_PROPERTY, "none");
-
-		    // Adding force suspend code to enable IVI to enter S3 after pressing sleep button
                     try {
-                        FileUtils.stringToFile("/sys/power/state", "mem");
-                    }
-                    catch (IOException e) {
-                        Slog.v(TAG, "IOException: " + e);
+                        FileUtils.stringToFile(ROLE_SYSFS, "host");
+                        Slog.i(TAG, "Switching to host mode before sleep...");
+                    } catch(IOException e) {
+                         Slog.v(TAG, "IOException: " + e);
                     }
                     break;
             }
@@ -1469,6 +1468,12 @@ public final class PowerManagerService extends SystemService
             }
         } finally {
             Trace.traceEnd(Trace.TRACE_TAG_POWER);
+	    // Adding force suspend code to enable IVI to enter S3 after pressing sleep button
+            try {
+                FileUtils.stringToFile("/sys/power/state", "mem");
+            } catch (IOException e) {
+                Slog.v(TAG, "IOException: " + e);
+            }
         }
         return true;
     }
@@ -1956,6 +1961,14 @@ public final class PowerManagerService extends SystemService
         String currUsbConfig = new String(SystemProperties.get(USB_CONFIG_PROPERTY, "none"));
         PowerManager mPowerManager = (PowerManager) mContext.getSystemService(Context.POWER_SERVICE);
 
+        
+        try {
+            FileUtils.stringToFile(ROLE_SYSFS, "device");
+            Slog.i(TAG, "Switching to host mode before sleep...");
+        } catch(IOException e) {
+            Slog.v(TAG, "IOException: " + e);
+        }
+
         if(currUsbConfig.equals("none") && (!mPowerManager.isInteractive())) {
                 SystemProperties.set(USB_CONFIG_PROPERTY, "adb");
 
-- 
2.21.0

